### PR TITLE
add identify call with test

### PR DESCRIPTION
**Pull request by lambtron**
Submitting this PR so we can chat about a few product decisions re: this feature.

image

Right now, the `identify` call to Slack works only if there is an `email` in the call.

Would that be too noisy in Slack? One thought @f2prateek had was to add a `traits` filter in the settings (similar to an `events` filter), where only `identify` calls that contain the `settings.traits` would be sent into Slack.

Anything else that i'm missing (tests, etc.)?

cc @reinpk if you have any comments

### CODE REVIEW

**Comment by reinpk** [comment]
The email filter seems a bit arbitrary from a user perspective. Sort of "too helpful"/magicky, yeah? Maybe we just send everything through and then later can release additional options for limiting it?

**Comment by f2prateek** [comment]
FWIW with event rules we might be able to send it through unconditionally and let users define their custom filtering behavior in a central space rather than in each integration. Though it will take a while to roll out.

File: lib/index.js [hunk 1]
```diff
@@ -21,49 +21,59 @@ var Slack = module.exports = integration('Slack')
...
-
-    return 'Anonymous user ' + track.anonymousId();
+Slack.prototype.identify = identify;
+Slack.prototype.track = track;
+
+/**
+ * Identify.
+ */
+
+function identify(identify, fn) {
```
Comment (hankim813) [inline lib/index.js]: should move all mapper logic into `lib/mapper.js` :)

Scope out some of our recently written server side integrations like facebook-offline-conversions or responsys

File: lib/index.js [hunk 2]
```diff
@@ -21,49 +21,59 @@ var Slack = module.exports = integration('Slack')
...
+  // `email` is needed to send to Slack.
+  if (!identify.email()) return fn('No email');
```
Comment (hankim813) [inline lib/index.js]: should put inside .ensure() block as seen here: https://github.com/segment-integrations/integration-facebook-offline-conversions/blob/master/lib/index.js#L27

File: lib/index.js [hunk 3]
```diff
@@ -21,49 +21,59 @@ var Slack = module.exports = integration('Slack')
...
+
+  var template = 'Identified new user, {{name}} at {{email}}. {{traits}}';
+  var compiled = handlebars.compile(template);
+  var traits = identify.json().traits;
```
Comment (hankim813) [inline lib/index.js]: can just use `identify.traits();`

File: lib/index.js [hunk 4]
```diff
@@ -21,49 +21,59 @@ var Slack = module.exports = integration('Slack')
...
+  var traitsText = '\n';
```
Comment (hankim813) [inline lib/index.js]: hmm this is confusing, why is this set to `\n`? Should be fine just as `var traitsText;`

File: lib/index.js [hunk 5]
```diff
@@ -85,3 +95,40 @@ function send(track, fn){
...
+ * 4. Check traits.lastName
+ * 5. Check traits.username
+ * 6. Check traits.email
+ * 7. Use 'User ' + userId
+ * 8. Use 'Anonymous user ' + anonymousId
+ *
+ * @return {string}
+ */
+
+function getName(message) {
```
Comment (hankim813) [inline lib/index.js]: we already have `msg.firstName()`, `msg.lastName()`, `msg.name()`, `msg.username()` that does alot of this lookup logic for you :)

File: lib/index.js (hunk 1 above)
Comment (f2prateek) [inline lib/index.js]: Not a required. e.g. https://github.com/segmentio/integration-direct/blob/master/lib/index.js

File: lib/index.js (hunk 5 above)
Comment (f2prateek) [inline lib/index.js]: I believe those don't look under `context.traits` - just `traits` so wouldn't work out of the box here.

File: lib/index.js (hunk 1 above)
Comment (hankim813) [inline lib/index.js]: not required but we don't want this kind of logic living in the index.js

Comment (f2prateek) [inline lib/index.js]: Agreed, but that's not a hard rule and this code base is already setup this way so it's not Andy's responsibility to change it.

**Comment by lambtron** [comment]
Thanks everyone for their input.

Just to be clear as to what is required for shipping this first iteration of this feature by the time our marketing campaign is launched so it's end-user-ready.

1. Remove `email` as filter.
2. only accept `newUser: true`? Or should we ignore that? This means that only all newly added `identify` calls _with_ the `newUser: true` will be sent to Slack. If we don't do `newUser: true`, then we'll have to add in a `traits` filter in the `settings`.
3. Should we build in the user custom templating for this iteration? Can we save this one for a future iteration? :)

**Comment by reinpk** [comment]
is *newUser* a spec'd trait?

File: lib/index.js (hunk 5 above)
Comment (hankim813) [inline lib/index.js]: you can just pass the facade object

File: lib/index.js (hunk 1 above)
Comment (lambtron) [inline lib/index.js]: i tried pulling it into a separate `mapper.js` file, but the try/catch block couldn't effectively propagate the error so we decided to keep the logic in `index.js`

File: lib/index.js (hunk 2 above)
Comment (lambtron) [inline lib/index.js]: I added an `.ensure()` block in the most recent one. but i'm not sure how to write tests for `.reject()`. any idea?

https://github.com/segment-integrations/integration-slack/pull/29/commits/a50836c2dfea3255277141e472ad205232fdd381

File: lib/index.js (hunk 3 above)
Comment (lambtron) [inline lib/index.js]: 👍

File: lib/index.js (hunk 4 above)
Comment (lambtron) [inline lib/index.js]: fixed this

File: lib/index.js (hunk 5 above)
Comment (lambtron) [inline lib/index.js]: the function `getName` is general enough to return `name` for both a `track` object and an `identify` object, which is why it has those funky `.proxy()` functions

**Comment by lambtron** [comment]
Two questions:

1:
How do i properly test for the `.ensure()` function's `this.reject()`?

2:
Currently, this PR will white list on `traits`, which means if you white list `.newUser`, then even when `newUser` is `undefined`, it'll still go through. would that be super noisy? 

I know we can add white listed _values_ with the keys, but it possible with `this.settings` to accept everything _except_ certain values (like `undefined` or `null`)? Or would it make sense to just do _values_ with the keys, e.g. "newUser = true;"?

**Comment by hankim813** [comment]
@lambtron 

1) basically do tests like this https://github.com/segment-integrations/integration-facebook-offline-conversions/blob/master/test/index.js#L54

2) I think that's fine. 

3) ehh unless you are on a strict deadline i'd encourage we support it, you just copy what `.track()` is doing

**Comment by lambtron** [comment]
@hankim813 added some more tests here. what's left is to update the metadata with a `settings.whiteListedTraits` and we should be set.

File: lib/index.js [hunk 6]
```diff
@@ -18,52 +20,69 @@ var Slack = module.exports = integration('Slack')
   .retries(1);
 
 /**
+ * Ensure.
+ */
+
+Slack.ensure(function(msg, settings) {
+  if (msg.type() === 'identify') {
+    if (settings.whiteListedTraits.length === 0) return this.reject('No white listed traits were set.');
```
Comment (hankim813) [inline lib/index.js]: I'd probably do

```
if (msg.type() === 'identify' && settings.whiteListedTraits && settings.whiteListedTraits.length) return this.reject('You must white list at least one trait for `.identify()` calls');
```

File: lib/index.js [hunk 7]
```diff
@@ -18,52 +20,69 @@ var Slack = module.exports = integration('Slack')
...
+    if (!includesKeys(msg.traits(), settings.whiteListedTraits)) return this.reject('Only identify calls that include `traits` in settings will be sent.');
```
Comment (hankim813) [inline lib/index.js]: `'Only identify calls that have match traits listed in the settings will be sent.'`

**Comment by reinpk** [comment]
🔥🚨💪

**Pull request merged**